src/borg/locking.py

```diff
@@ -124,30 +125,50 @@ def acquire(self, timeout=None, sleep=None):
             timeout = self.timeout
         if sleep is None:
             sleep = self.sleep
-        timer = TimeoutTimer(timeout, sleep).start()
-        while True:
-            try:
-                os.mkdir(self.path)
-            except FileExistsError:  # already locked
-                if self.by_me():
+        parent_path, base_name = os.path.split(self.path)
+        unique_base_name = os.path.basename(self.unique_name)
+        temp_path = None
+        try:
+            temp_path = tempfile.mkdtemp(".tmp", base_name + '.', parent_path)
+            temp_unique_name = os.path.join(temp_path, unique_base_name)
+            with open(temp_unique_name, "wb"):
+                pass
+        except OSError as err:
+            raise LockFailed(self.path, str(err)) from None
+        else:
+            timer = TimeoutTimer(timeout, sleep).start()
+            while True:
+                try:
+                    os.rename(temp_path, self.path)
+                except OSError:  # already locked
+                    if self.by_me():
+                        return self
+                    self.kill_stale_lock()
+                    if timer.timed_out_or_sleep():
+                        raise LockTimeout(self.path)
+                else:
+                    temp_path = None
                     return self
-                self.kill_stale_lock()
-                if timer.timed_out_or_sleep():
-                    raise LockTimeout(self.path)
-            except OSError as err:
-                raise LockFailed(self.path, str(err)) from None
-            else:
-                with open(self.unique_name, "wb"):
-                    pass
-                return self
+        finally:
+            if temp_path is not None:  # Renaming failed for any reason, so temp_dir still exists and should be cleaned up anyway.
+                os.unlink(temp_unique_name)
+                os.rmdir(temp_path)
```

Comment:
if the open-wb fails, there is no temp_unique_name file, so guess we better put the first line into a `try: ... except: pass`

Reply:
... and maybe also the second line into a separate handler (or do we ever want to fail while trying to clean up?)

Reply:
done

Reply:
Probably you're right. I remember I thought about it, and in my opinion it wasn't necessary, because if such kind of things happen, it would have already happened before. But basically, everything can happen during I/O at any time...